### PR TITLE
added default tslint.json, implemented local copy and rename

### PR DESCRIPTION
**Pull request by jmartin84**

### CODE REVIEW

File: src/adapters/local.ts [hunk 1]
```diff
@@ -42,11 +42,24 @@ export default class LocalAdapter implements IAdapter {
...
+        const readableStream = fs.createReadStream(this.fullPath(oldPath));
+        const writeableStream = fs.createWriteStream(this.fullPath(clonedPath));
+
+        // we use streams here to ensure that copy will always work regardless of file size
+        await new Bluebird((resolve, reject) => {
+            readableStream
+                .pipe(writeableStream)
+                .on('error', () => reject(false))
+                .on('finish', () => resolve(true));
+        });
```
Comment (jmartin84) [inline src/adapters/local.ts]: there has to be a better way to do this.

Comment (CWSpear) [inline src/adapters/local.ts]: Don't know where my other comment went.

Can we do something along the lines of:

```js
return this.writeStream(newPath, (await this.readStream).stream);
```

File: src/adapters/local.ts [hunk 2]
```diff
@@ -42,11 +42,24 @@ export default class LocalAdapter implements IAdapter {
     }
 
     async rename(oldPath: string, newPath: string): Promise<boolean> {
-        throw new Error('rename NYI');
+        await fs.renameAsync(this.fullPath(oldPath), this.fullPath(newPath));
+
+        return true;
```
Comment (CWSpear) [inline src/adapters/local.ts]: so yeah, we need to go through and decide what we want stuff to return. Just `true` or `Metadata`?

And so you know, the return type is `Promise<boolean>`, where the stuff in the `<>` is what the promise resolves to. So we're considering changing it to `Promise<IMetadata>`. I don't like the `ISomething` pattern for interfaces, but it's apparently pretty standard in TypeScript, and I want people familiar with TypeScript to be able to use these interfaces.

**Comment by CWSpear** [comment]
I don't see any changes to the `package.json`. Is there not a `tslint` that should be there? And maybe a `lint` script?

**Comment by CWSpear** [comment]
You can create a new PR with the linting stuff (including the needed changes in `package.json`), but I implemented the `copy` and `rename` with `fs-extra`.

File: src/adapters/local.ts (hunk 1 above)
Comment (jmartin84) [inline src/adapters/local.ts]: sure

File: src/adapters/local.ts (hunk 2 above)
Comment (jmartin84) [inline src/adapters/local.ts]: yeah I know, its just a generic.

ISomething comes from .net - I personally prefer it.

so are we doing a boolean or metadata?

**Comment by jmartin84** [comment]
whats fs-extra?

**Comment by CWSpear** [comment]
`fs-extra` adds on some extra features and falls back to `fs` for other things (I think `rename` is one of them). I used it for a few other things. I am actually using `fs-extra-promise` because not all of the methods were `Bluebird.promisifyAll`-able.

File: src/adapters/local.ts (hunk 2 above)
Comment (CWSpear) [inline src/adapters/local.ts]: Metadata. See the updated code, tests and interfaces.

**Pull request closed without merging**